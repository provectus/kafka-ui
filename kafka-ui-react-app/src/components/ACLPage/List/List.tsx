import React from 'react';
import { ColumnDef } from '@tanstack/react-table';
import { useTheme } from 'styled-components';
import PageHeading from 'components/common/PageHeading/PageHeading';
import Table from 'components/common/NewTable';
import DeleteIcon from 'components/common/Icons/DeleteIcon';
import { useConfirm } from 'lib/hooks/useConfirm';
import useAppParams from 'lib/hooks/useAppParams';
import { useAcls, useDeleteAcl } from 'lib/hooks/api/acl';
import { ClusterName } from 'redux/interfaces';
import {
  KafkaAcl,
  KafkaAclNamePatternType,
  KafkaAclPermissionEnum,
} from 'generated-sources';

import * as S from './List.styled';

const ACList: React.FC = () => {
  const { clusterName } = useAppParams<{ clusterName: ClusterName }>();
  const theme = useTheme();
  const { data: aclList } = useAcls(clusterName);
  const { deleteResource } = useDeleteAcl(clusterName);
  const modal = useConfirm(true);

  const [rowId, setRowId] = React.useState('');

  const onDeleteClick = (acl: KafkaAcl | null) => {
    if (acl) {
      modal('Are you sure want to delete this ACL record?', () =>
        deleteResource(acl)
      );
    }
  };

  const columns = React.useMemo<ColumnDef<KafkaAcl>[]>(
    () => [
      {
        header: 'Principal',
        accessorKey: 'principal',
        size: 257,
      },
      {
        header: 'Resource',
        accessorKey: 'resourceType',
        // eslint-disable-next-line react/no-unstable-nested-components
        cell: ({ getValue }) => (
          <S.EnumCell>{getValue<string>().toLowerCase()}</S.EnumCell>
        ),
        size: 145,
      },
      {
        header: 'Pattern',
        accessorKey: 'resourceName',
        // eslint-disable-next-line react/no-unstable-nested-components
        cell: ({ getValue, row }) => {
          let chipType;
          if (
            row.original.namePatternType === KafkaAclNamePatternType.PREFIXED
          ) {
            chipType = 'default';
          }

          if (
            row.original.namePatternType === KafkaAclNamePatternType.LITERAL
          ) {
            chipType = 'secondary';
          }
          return (
            <S.PatternCell>
              {getValue<string>()}
              {chipType ? (
                <S.Chip chipType={chipType}>
                  {row.original.namePatternType.toLowerCase()}
                </S.Chip>
              ) : null}
            </S.PatternCell>
          );
        },
        size: 257,
      },
      {
        header: 'Host',
        accessorKey: 'host',
        size: 257,
      },
      {
        header: 'Operation',
        accessorKey: 'operation',
        // eslint-disable-next-line react/no-unstable-nested-components
        cell: ({ getValue }) => (
          <S.EnumCell>{getValue<string>().toLowerCase()}</S.EnumCell>
        ),
        size: 121,
      },
      {
        header: 'Permission',
        accessorKey: 'permission',
        // eslint-disable-next-line react/no-unstable-nested-components
        cell: ({ getValue }) => (
          <S.Chip
            chipType={
              getValue<string>() === KafkaAclPermissionEnum.ALLOW
                ? 'success'
                : 'danger'
            }
          >
            {getValue<string>().toLowerCase()}
          </S.Chip>
        ),
        size: 111,
      },
      {
        id: 'delete',
        // eslint-disable-next-line react/no-unstable-nested-components
        cell: ({ row }) => {
          return (
            <S.DeleteCell onClick={() => onDeleteClick(row.original)}>
              <DeleteIcon
                fill={
                  rowId === row.id ? theme.acl.table.deleteIcon : 'transparent'
                }
              />
            </S.DeleteCell>
          );
        },
        size: 76,
      },
    ],
    [rowId]
  );

  const onRowHover = (value: unknown) => {
    if (value && typeof value === 'object' && 'id' in value) {
      setRowId(value.id as string);
    }
  };

  return (
    <>
      <PageHeading text="Access Control List" />
      <Table
        columns={columns}
        data={aclList ?? []}
        emptyMessage="No ACL items found"
        onRowHover={onRowHover}
        onMouseLeave={() => setRowId('')}
      />
    </>
  );
};

export default ACList;
