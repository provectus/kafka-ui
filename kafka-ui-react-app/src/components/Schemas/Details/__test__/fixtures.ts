import { SchemaSubject, SchemaType } from 'generated-sources';
import {
  schemaVersion1,
  schemaVersion2,
  schemaVersionWithNonAsciiChars,
} from 'redux/reducers/schemas/__test__/fixtures';

export const versionPayload = [
  schemaVersion1,
  schemaVersion2,
  schemaVersionWithNonAsciiChars,
];
export const versionEmptyPayload = [];

export const jsonSchema: SchemaSubject = {
  subject: 'test',
  version: '15',
  id: 1,
  schema:
    '{"type":"record","name":"MyRecord1","namespace":"com.mycompany","fields":[{"name":"id","type":"long"}]}',
  compatibilityLevel: 'BACKWARD',
  schemaType: SchemaType.JSON,
};

export const protoSchema: SchemaSubject = {
  subject: 'test_proto',
  version: '1',
  id: 2,
  schema:
    'syntax = "proto3";\npackage com.indeed;\n\nmessage MyRecord {\n  int32 id = 1;\n  string name = 2;\n}\n',
  compatibilityLevel: 'BACKWARD',
  schemaType: SchemaType.PROTOBUF,
};
