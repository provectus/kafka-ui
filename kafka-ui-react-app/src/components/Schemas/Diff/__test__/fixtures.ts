import { SchemaSubject, SchemaType } from 'generated-sources';

export const versions: SchemaSubject[] = [
  {
    subject: 'test',
    version: '3',
    id: 3,
    schema:
      'syntax = "proto3";\npackage com.indeed;\n\nmessage MyRecord {\n  int32 id = 1;\n  string name = 2;\n}\n',
    compatibilityLevel: 'BACKWARD',
    schemaType: SchemaType.PROTOBUF,
  },
  {
    subject: 'test',
    version: '2',
    id: 2,
    schema:
      '{"type":"record","name":"MyRecord2","namespace":"com.mycompany","fields":[{"name":"id","type":"long"}]}',
    compatibilityLevel: 'BACKWARD',
    schemaType: SchemaType.JSON,
  },
  {
    subject: 'test',
    version: '1',
    id: 1,
    schema:
      '{"type":"record","name":"MyRecord1","namespace":"com.mycompany","fields":[{"name":"id","type":"long"}]}',
    compatibilityLevel: 'BACKWARD',
    schemaType: SchemaType.JSON,
  },
];
