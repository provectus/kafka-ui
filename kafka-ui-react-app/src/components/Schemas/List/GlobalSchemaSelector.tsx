import ConfirmationModal from 'components/common/ConfirmationModal/ConfirmationModal';
import PageLoader from 'components/common/PageLoader/PageLoader';
import { CompatibilityLevelCompatibilityEnum } from 'generated-sources';
import React from 'react';
import { useForm } from 'react-hook-form';
import { useParams } from 'react-router-dom';
import { ClusterName } from 'redux/interfaces';

export interface GlobalSchemaSelectorProps {
  globalSchemaCompatibilityLevel?: CompatibilityLevelCompatibilityEnum;
  updateGlobalSchemaCompatibilityLevel: (
    clusterName: ClusterName,
    compatibilityLevel: CompatibilityLevelCompatibilityEnum
  ) => Promise<void>;
}

const GlobalSchemaSelector = ({
  globalSchemaCompatibilityLevel,
  updateGlobalSchemaCompatibilityLevel,
}: GlobalSchemaSelectorProps) => {
  const { clusterName } = useParams<{ clusterName: string }>();

  const {
    register,
    handleSubmit,
    formState: { isSubmitting },
  } = useForm();

  const [
    isUpdateCompatibilityConfirmationVisible,
    setUpdateCompatibilityConfirmationVisible,
  ] = React.useState(false);

  const onCompatibilityLevelUpdate = async ({
    compatibilityLevel,
  }: {
    compatibilityLevel: CompatibilityLevelCompatibilityEnum;
  }) => {
    await updateGlobalSchemaCompatibilityLevel(clusterName, compatibilityLevel);
    setUpdateCompatibilityConfirmationVisible(false);
  };

  return (
    <div className="level-item">
      <h5 className="is-5 mr-2">Global Compatibility Level: </h5>
      <div className="select mr-2">
        <select
          name="compatibilityLevel"
          defaultValue={globalSchemaCompatibilityLevel}
          ref={register()}
          onChange={() => setUpdateCompatibilityConfirmationVisible(true)}
        >
          {Object.keys(CompatibilityLevelCompatibilityEnum).map(
            (level: string) => (
              <option key={level} value={level}>
                {level}
              </option>
            )
          )}
        </select>
      </div>
      <ConfirmationModal
        isOpen={isUpdateCompatibilityConfirmationVisible}
        onCancel={() => setUpdateCompatibilityConfirmationVisible(false)}
        onConfirm={handleSubmit(onCompatibilityLevelUpdate)}
      >
        {isSubmitting ? (
          <PageLoader />
        ) : (
          `Are you sure you want to update the global compatibility level?
                  This may affect the compatibility levels of the schemas.`
        )}
      </ConfirmationModal>
    </div>
  );
};

export default GlobalSchemaSelector;
