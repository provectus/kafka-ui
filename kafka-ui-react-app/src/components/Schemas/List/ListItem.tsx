import React from 'react';
import { SchemaSubject } from 'generated-sources';
import { NavLink } from 'react-router-dom';

export interface ListItemProps {
  subject: SchemaSubject;
}

const ListItem: React.FC<ListItemProps> = ({
  subject: { subject, version, compatibilityLevel },
}) => {
  return (
    <tr>
      <td>
        <NavLink
          exact
          to={`schemas/${subject}/latest`}
          activeClassName="is-active"
          className="title is-6"
        >
          {subject}
        </NavLink>
      </td>
      <td>{version}</td>
      <td>
        <span className="tag is-link">{compatibilityLevel}</span>
      </td>
    </tr>
  );
};

export default ListItem;
