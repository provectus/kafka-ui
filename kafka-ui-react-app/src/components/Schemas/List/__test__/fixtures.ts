import { SchemaSubject } from 'generated-sources';

export const schemas: SchemaSubject[] = [
  {
    subject: 'test',
    version: '1',
    id: 1,
    schema:
      '{"type":"record","name":"MyRecord1","namespace":"com.mycompany","fields":[{"name":"id","type":"long"}]}',
    compatibilityLevel: 'BACKWARD',
  },
  {
    subject: 'test2',
    version: '1',
    id: 1,
    schema:
      '{"type":"record","name":"MyRecord2","namespace":"com.mycompany","fields":[{"name":"id","type":"long"}]}',
    compatibilityLevel: 'BACKWARD',
  },
  {
    subject: 'test3',
    version: '1',
    id: 1,
    schema:
      '{"type":"record","name":"MyRecord3","namespace":"com.mycompany","fields":[{"name":"id","type":"long"}]}',
    compatibilityLevel: 'BACKWARD',
  },
];
