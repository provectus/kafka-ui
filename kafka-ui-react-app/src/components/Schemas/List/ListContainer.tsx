import { connect } from 'react-redux';
import { RootState } from 'redux/interfaces';
import { getSchemaList } from 'redux/reducers/schemas/selectors';
import List from './List';

const mapStateToProps = (state: RootState) => ({
  schemas: getSchemaList(state),
});

export default connect(mapStateToProps)(List);
