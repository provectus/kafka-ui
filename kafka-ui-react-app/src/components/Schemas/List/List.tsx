import React from 'react';
import { SchemaSubject } from 'generated-sources';
import Breadcrumb from '../../common/Breadcrumb/Breadcrumb';
import ListItem from './ListItem';

export interface ListProps {
  schemas: SchemaSubject[];
}

const List: React.FC<ListProps> = ({ schemas }) => {
  return (
    <div className="section">
      <Breadcrumb>Schema Registry</Breadcrumb>
      <div className="box">
        <table className="table is-striped is-fullwidth">
          <thead>
            <tr>
              <th>Schema Name</th>
              <th>Version</th>
              <th>Compatibility</th>
            </tr>
          </thead>
          <tbody>
            {schemas.map((subject) => (
              <ListItem key={subject.id} subject={subject} />
            ))}
          </tbody>
        </table>
      </div>
    </div>
  );
};

export default List;
