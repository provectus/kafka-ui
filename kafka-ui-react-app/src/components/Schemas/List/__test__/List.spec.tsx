import React from 'react';
import List from 'components/Schemas/List/List';
import { render, WithRoute } from 'lib/testHelpers';
import { clusterSchemaPath, clusterSchemasPath } from 'lib/paths';
import { act, screen } from '@testing-library/react';
import {
  schemasFulfilledState,
  schemasInitialState,
  schemaVersion1,
  schemaVersion2,
} from 'redux/reducers/schemas/__test__/fixtures';
import ClusterContext, {
  ContextProps,
  initialValue as contextInitialValue,
} from 'components/contexts/ClusterContext';
import { RootState } from 'redux/interfaces';
import fetchMock from 'fetch-mock';
import userEvent from '@testing-library/user-event';

import { schemasPayload, schemasEmptyPayload } from './fixtures';

const mockedUsedNavigate = jest.fn();

jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useNavigate: () => mockedUsedNavigate,
}));

const clusterName = 'testClusterName';
const schemasAPIUrl = `/api/clusters/${clusterName}/schemas?page=1&perPage=25`;
const schemasAPICompabilityUrl = `/api/clusters/${clusterName}/schemas/compatibility`;
const renderComponent = (
  initialState: RootState['schemas'] = schemasInitialState,
  context: ContextProps = contextInitialValue
) =>
  render(
    <WithRoute path={clusterSchemasPath()}>
      <ClusterContext.Provider value={context}>
        <List />
      </ClusterContext.Provider>
    </WithRoute>,
    {
      initialEntries: [clusterSchemasPath(clusterName)],
      preloadedState: {
        schemas: initialState,
      },
    }
  );

describe('List', () => {
  afterEach(() => {
    fetchMock.reset();
  });

  describe('fetch error', () => {
    it('shows progressbar', async () => {
      const fetchSchemasMock = fetchMock.getOnce(schemasAPIUrl, 404);
      const fetchCompabilityMock = fetchMock.getOnce(
        schemasAPICompabilityUrl,
        404
      );
      await act(() => {
        renderComponent();
      });
      expect(fetchSchemasMock.called()).toBeTruthy();
      expect(fetchCompabilityMock.called()).toBeTruthy();
      expect(screen.getByRole('progressbar')).toBeInTheDocument();
    });
  });

  describe('fetch success', () => {
    describe('responded without schemas', () => {
      beforeEach(async () => {
        const fetchSchemasMock = fetchMock.getOnce(
          schemasAPIUrl,
          schemasEmptyPayload
        );
        const fetchCompabilityMock = fetchMock.getOnce(
          schemasAPICompabilityUrl,
          200
        );
        await act(() => {
          renderComponent();
        });
        expect(fetchSchemasMock.called()).toBeTruthy();
        expect(fetchCompabilityMock.called()).toBeTruthy();
      });
      it('renders empty table', () => {
        expect(screen.getByText('No schemas found')).toBeInTheDocument();
      });
    });
    describe('responded with schemas', () => {
      beforeEach(async () => {
        const fetchSchemasMock = fetchMock.getOnce(
          schemasAPIUrl,
          schemasPayload
        );
        const fetchCompabilityMock = fetchMock.getOnce(
          schemasAPICompabilityUrl,
          200
        );
        await act(() => {
          renderComponent(schemasFulfilledState);
        });
        expect(fetchSchemasMock.called()).toBeTruthy();
        expect(fetchCompabilityMock.called()).toBeTruthy();
      });
      it('renders list', () => {
        expect(screen.getByText(schemaVersion1.subject)).toBeInTheDocument();
        expect(screen.getByText(schemaVersion2.subject)).toBeInTheDocument();
      });
      it('handles onRowClick', async () => {
        const { id, schemaType, subject, version, compatibilityLevel } =
          schemaVersion2;
        const row = screen.getByRole('row', {
          name: `${subject} ${id} ${schemaType} ${version} ${compatibilityLevel}`,
        });
        expect(row).toBeInTheDocument();
        await userEvent.click(row);
        expect(mockedUsedNavigate).toHaveBeenCalledWith(
          clusterSchemaPath(clusterName, subject)
        );
      });
    });

    describe('responded with readonly cluster schemas', () => {
      beforeEach(async () => {
        const fetchSchemasMock = fetchMock.getOnce(
          schemasAPIUrl,
          schemasPayload
        );
        fetchMock.getOnce(schemasAPICompabilityUrl, 200);
        await act(() => {
          renderComponent(schemasFulfilledState, {
            ...contextInitialValue,
            isReadOnly: true,
          });
        });
        expect(fetchSchemasMock.called()).toBeTruthy();
      });
      it('does not render Create Schema button', () => {
        expect(screen.queryByText('Create Schema')).not.toBeInTheDocument();
      });
    });
  });
});
