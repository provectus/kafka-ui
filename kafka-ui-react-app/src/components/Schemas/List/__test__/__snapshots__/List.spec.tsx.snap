// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`List View with schemas matches snapshot 1`] = `
<div
  className="section"
>
  <Breadcrumb>
    Schema Registry
  </Breadcrumb>
  <div
    className="box"
  >
    <table
      className="table is-striped is-fullwidth"
    >
      <thead>
        <tr>
          <th>
            Schema Name
          </th>
          <th>
            Version
          </th>
          <th>
            Compatibility
          </th>
        </tr>
      </thead>
      <tbody>
        <ListItem
          key="1"
          subject={
            Object {
              "compatibilityLevel": "BACKWARD",
              "id": 1,
              "schema": "{\\"type\\":\\"record\\",\\"name\\":\\"MyRecord1\\",\\"namespace\\":\\"com.mycompany\\",\\"fields\\":[{\\"name\\":\\"id\\",\\"type\\":\\"long\\"}]}",
              "subject": "test",
              "version": "1",
            }
          }
        />
        <ListItem
          key="1"
          subject={
            Object {
              "compatibilityLevel": "BACKWARD",
              "id": 1,
              "schema": "{\\"type\\":\\"record\\",\\"name\\":\\"MyRecord2\\",\\"namespace\\":\\"com.mycompany\\",\\"fields\\":[{\\"name\\":\\"id\\",\\"type\\":\\"long\\"}]}",
              "subject": "test2",
              "version": "1",
            }
          }
        />
        <ListItem
          key="1"
          subject={
            Object {
              "compatibilityLevel": "BACKWARD",
              "id": 1,
              "schema": "{\\"type\\":\\"record\\",\\"name\\":\\"MyRecord3\\",\\"namespace\\":\\"com.mycompany\\",\\"fields\\":[{\\"name\\":\\"id\\",\\"type\\":\\"long\\"}]}",
              "subject": "test3",
              "version": "1",
            }
          }
        />
      </tbody>
    </table>
  </div>
</div>
`;

exports[`List View without schemas matches snapshot 1`] = `
<div
  className="section"
>
  <Breadcrumb>
    Schema Registry
  </Breadcrumb>
  <div
    className="box"
  >
    <table
      className="table is-striped is-fullwidth"
    >
      <thead>
        <tr>
          <th>
            Schema Name
          </th>
          <th>
            Version
          </th>
          <th>
            Compatibility
          </th>
        </tr>
      </thead>
      <tbody />
    </table>
  </div>
</div>
`;
